keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java

```diff
@@ -21,50 +26,55 @@
 import static java.util.stream.Collectors.toMap;
 
 
-public class CategoricalVertex<T> extends Vertex<Tensor<T>> implements Probabilistic<Tensor<T>> {
+public class CategoricalVertex<T, TENSOR extends Tensor<T>> extends Vertex<TENSOR> implements Probabilistic<TENSOR> {
 
     private final Map<T, DoubleVertex> selectableValues;
 
-    public static <T> CategoricalVertex<T> of(Map<T, Double> selectableValues) {
-        return new CategoricalVertex<>(defensiveCopy(selectableValues));
+    public static <T, TENSOR extends Tensor<T>> CategoricalVertex<T, TENSOR> of(Map<T, Double> selectableValues) {
+        return new CategoricalVertex<>(toDoubleVertices(selectableValues));
     }
 
-    public static <T> CategoricalVertex<T> of(DirichletVertex vertex, List<T> categories) {
+    private static <T> Map<T, DoubleVertex> toDoubleVertices(Map<T, Double> selectableValues) {
+        Map<T, DoubleVertex> copy = new LinkedHashMap<>();
+        for (Map.Entry<T, Double> entry : selectableValues.entrySet()) {
+            copy.put(entry.getKey(), ConstantVertex.of(entry.getValue()));
+        }
+        return copy;
```

Comment:
return selectableValues.entrySet().stream()
            .collect(toMap(
                Map.Entry::getKey, 
                e -> ConstantVertex.of(e.getValue())
            ));